File: app/src/main/java/com/scatl/uestcbbs/module/home/view/CollectionListFragment.java
package com.scatl.uestcbbs.module.home.view;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.animation.AnimationUtils;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.scatl.uestcbbs.R;
import com.scatl.uestcbbs.base.BaseFragment;
import com.scatl.uestcbbs.base.BasePresenter;
import com.scatl.uestcbbs.callback.OnRefresh;
import com.scatl.uestcbbs.custom.MyLinearLayoutManger;
import com.scatl.uestcbbs.entity.CollectionListBean;
import com.scatl.uestcbbs.module.collection.view.CollectionActivity;
import com.scatl.uestcbbs.module.home.adapter.CollectionListAdapter;
import com.scatl.uestcbbs.module.home.adapter.MyCollectionListAdapter;
import com.scatl.uestcbbs.module.home.presenter.CollectionListPresenter;
import com.scatl.uestcbbs.module.post.view.PostDetailActivity;
import com.scatl.uestcbbs.module.user.view.UserDetailActivity;
import com.scatl.uestcbbs.util.Constant;
import com.scatl.uestcbbs.util.RefreshUtil;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.constant.RefreshState;

import java.util.ArrayList;
import java.util.List;


public class CollectionListFragment extends BaseFragment implements CollectionListView {

    private SmartRefreshLayout refreshLayout;
    private RecyclerView recyclerView;
    private CollectionListAdapter collectionListAdapter;
    private ProgressBar progressBar;
    private TextView hint;

    private View myCollectionListView;
    private RecyclerView myCollectionListRv;
    private TextView myCollectionListTitle;
    private View titleLayout;
    private MyCollectionListAdapter myCollectionListAdapter;

    private CollectionListPresenter collectionListPresenter;

    private int page = 1;

    public static CollectionListFragment getInstance(Bundle bundle) {
        CollectionListFragment collectionListFragment = new CollectionListFragment();
        collectionListFragment.setArguments(bundle);
        return collectionListFragment;
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.fragment_collection_list;
    }

    @Override
    protected void findView() {
        refreshLayout = view.findViewById(R.id.tao_tie_collection_refresh);
        recyclerView = view.findViewById(R.id.tao_tie_collection_rv);
        progressBar = view.findViewById(R.id.tao_tie_progressbar);
        hint = view.findViewById(R.id.tao_tie_hint);

        myCollectionListView = LayoutInflater.from(mActivity).inflate(R.layout.view_collection_list_my, new LinearLayout(mActivity));
        myCollectionListRv = myCollectionListView.findViewById(R.id.view_collection_list_my_rv);
        myCollectionListTitle = myCollectionListView.findViewById(R.id.view_collection_list_my_title);
        titleLayout = myCollectionListView.findViewById(R.id.view_collection_list_my_title_layout);
    }

    @Override
    protected void initView() {

        collectionListPresenter = (CollectionListPresenter) presenter;

        //我的专辑
        myCollectionListAdapter = new MyCollectionListAdapter(R.layout.item_collection_list_my);
        MyLinearLayoutManger myLinearLayoutManger = new MyLinearLayoutManger(mActivity);
        myLinearLayoutManger.setOrientation(LinearLayoutManager.HORIZONTAL);
        myCollectionListRv.setLayoutManager(myLinearLayoutManger);
        myCollectionListRv.setAdapter(myCollectionListAdapter);

        //所有专辑
        collectionListAdapter = new CollectionListAdapter(R.layout.item_collection_list);
        collectionListAdapter.addHeaderView(myCollectionListView, 0);
        recyclerView.setLayoutManager(new MyLinearLayoutManger(mActivity));
        recyclerView.setLayoutAnimation(AnimationUtils.loadLayoutAnimation(mActivity, R.anim.layout_animation_scale_in));
        recyclerView.setAdapter(collectionListAdapter);
    }

    @Override
    protected void lazyLoad() {
        refreshLayout.autoRefresh(0, 300, 1, false);
    }

    @Override
    protected BasePresenter initPresenter() {
        return new CollectionListPresenter();
    }

    @Override
    protected void setOnItemClickListener() {
        collectionListAdapter.setOnItemChildClickListener((adapter, view1, position) -> {
            if (view1.getId() == R.id.item_tao_tie_collection_latest_post) {
                Intent intent = new Intent(mActivity, PostDetailActivity.class);
                intent.putExtra(Constant.IntentKey.TOPIC_ID, collectionListAdapter.getData().get(position).latestPostId);
                startActivity(intent);
            }

            if (view1.getId() == R.id.item_tao_tie_collection_user_avatar) {
                Intent intent = new Intent(mActivity, UserDetailActivity.class);
                intent.putExtra(Constant.IntentKey.USER_ID, collectionListAdapter.getData().get(position).authorId);
                startActivity(intent);
            }
        });

        collectionListAdapter.setOnItemClickListener((adapter, view1, position) -> {
            if (view1.getId() == R.id.item_tao_tie_collection_card_view) {
                Intent intent = new Intent(mActivity, CollectionActivity.class);
                intent.putExtra(Constant.IntentKey.COLLECTION_ID, collectionListAdapter.getData().get(position).collectionId);
                startActivity(intent);
            }
        });

        myCollectionListAdapter.setOnItemChildClickListener((adapter, view1, position) -> {
            if (view1.getId() == R.id.item_collection_list_my_user_avatar) {
                Intent intent = new Intent(mActivity, UserDetailActivity.class);
                intent.putExtra(Constant.IntentKey.USER_ID, myCollectionListAdapter.getData().get(position).authorId);
                startActivity(intent);
            }
        });

        myCollectionListAdapter.setOnItemClickListener((adapter, view1, position) -> {
            if (view1.getId() == R.id.item_collection_list_my_card_view) {
                Intent intent = new Intent(mActivity, CollectionActivity.class);
                intent.putExtra(Constant.IntentKey.COLLECTION_ID, myCollectionListAdapter.getData().get(position).collectionId);
                startActivity(intent);
            }
        });
    }

    @Override
    protected void setOnRefreshListener() {
        RefreshUtil.setOnRefreshListener(mActivity, refreshLayout, new OnRefresh() {
            @Override
            public void onRefresh(RefreshLayout refreshLayout) {
                page = 1;
                collectionListPresenter.getCollectionList(page, "all");//todo
                collectionListPresenter.getMyCollection();
            }

            @Override
            public void onLoadMore(RefreshLayout refreshLayout) {
                collectionListPresenter.getCollectionList(page,"all");//todo
            }
        });
    }

    @Override
    public void onGetCollectionListSuccess(List<CollectionListBean> collectionBeans, boolean hasNext) {
        progressBar.setVisibility(View.GONE);
        hint.setText("");

        if (page == 1) {
            collectionListAdapter.setNewData(collectionBeans);
            recyclerView.scheduleLayoutAnimation();
        }
        if (page != 1) collectionListAdapter.addData(collectionBeans);


        if (refreshLayout.getState() == RefreshState.Refreshing) {
            page = 2;
            if (hasNext) {
                refreshLayout.finishRefresh();
            } else {
                refreshLayout.finishRefreshWithNoMoreData();
            }
        }
        if (refreshLayout.getState() == RefreshState.Loading) {
            if (hasNext) {
                page = page + 1;
                refreshLayout.finishLoadMore(true);
            } else {
                refreshLayout.finishLoadMoreWithNoMoreData();
            }
        }
    }

    @Override
    public void onGetCollectionListError(String msg) {
        if (refreshLayout.getState() == RefreshState.Refreshing) {
            refreshLayout.finishRefresh();
        }
        if (refreshLayout.getState() == RefreshState.Loading) {
            refreshLayout.finishLoadMore(false);
        }
        hint.setText(msg);
    }

    @Override
    public void onGetMyCollectionSuccess(List<CollectionListBean> collectionListBeans) {
        titleLayout.setVisibility(View.VISIBLE);
        myCollectionListAdapter.setNewData(collectionListBeans);
        myCollectionListTitle.setText("我的专辑（" + collectionListBeans.size() + "）");
    }

    @Override
    public void onGetMyCollectionError(String msg) {
        titleLayout.setVisibility(View.GONE);
        myCollectionListAdapter.setNewData(new ArrayList<>());
    }
}

File: app/src/main/java/com/scatl/uestcbbs/module/collection/view/CollectionListFragment.kt
package com.scatl.uestcbbs.module.collection.view

import android.content.Intent
import android.os.Bundle
import android.view.HapticFeedbackConstants
import android.view.View
import android.view.animation.AnimationUtils
import androidx.recyclerview.widget.RecyclerView
import com.scatl.uestcbbs.R
import com.scatl.uestcbbs.annotation.ToastType
import com.scatl.uestcbbs.base.BaseEvent
import com.scatl.uestcbbs.base.BaseVBFragment
import com.scatl.uestcbbs.databinding.FragmentCollectionListBinding
import com.scatl.uestcbbs.entity.CollectionListBean
import com.scatl.uestcbbs.module.collection.adapter.CollectionListAdapter
import com.scatl.uestcbbs.module.collection.presenter.CollectionListPresenter
import com.scatl.uestcbbs.module.post.view.NewPostDetailActivity
import com.scatl.uestcbbs.module.user.view.UserDetailActivity
import com.scatl.uestcbbs.util.Constant
import com.scatl.uestcbbs.util.showToast
import com.scwang.smart.refresh.layout.api.RefreshLayout
import org.greenrobot.eventbus.EventBus

/**
 * Created by sca_tl at 2023/5/5 11:41
 */
class CollectionListFragment: BaseVBFragment<CollectionListPresenter, CollectionListView, FragmentCollectionListBinding>(), CollectionListView {

    private lateinit var collectionListAdapter: CollectionListAdapter
    private var mType = TYPE_ALL
    private var mOrder = ORDER_FOLLOW_NUM
    private var mPage = 1

    companion object {
        const val TYPE_MINE = "my"
        const val TYPE_RECOMMEND = "recommend"
        const val TYPE_ALL = "all"

        const val ORDER_CREATE_TIME = "createtime"
        const val ORDER_FOLLOW_NUM = "follownum"
        const val ORDER_THREAD_NUM = "threadnum"
        const val ORDER_COMMENT_NUM = "commentnum"

        fun getInstance(bundle: Bundle?) = CollectionListFragment().apply { arguments = bundle }
    }

    override fun getViewBinding() = FragmentCollectionListBinding.inflate(layoutInflater)

    override fun initPresenter() = CollectionListPresenter()

    override fun getBundle(bundle: Bundle?) {
        super.getBundle(bundle)
        mType = bundle?.getString(Constant.IntentKey.TYPE, TYPE_ALL)?: TYPE_ALL
    }

    override fun initView() {
        super.initView()
        collectionListAdapter = CollectionListAdapter(R.layout.item_collection_list)

        if (mType == TYPE_ALL) {
            mBinding.chipGroup.visibility = View.VISIBLE
            mBinding.commentNumSortBtn.setOnClickListener(this)
            mBinding.createTimeSortBtn.setOnClickListener(this)
            mBinding.followNumSortBtn.setOnClickListener(this)
            mBinding.postNumSortBtn.setOnClickListener(this)
            mBinding.followNumSortBtn.isChecked = true
        } else {
            mBinding.chipGroup.visibility = View.GONE
        }

        mBinding.recyclerView.apply {
            adapter = collectionListAdapter
            layoutAnimation = AnimationUtils.loadLayoutAnimation(context, R.anim.layout_animation_from_top)
            addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    EventBus.getDefault().post(BaseEvent(BaseEvent.EventCode.HOME_NAVIGATION_HIDE, dy > 0))
                }
            })
        }
        mBinding.statusView.loading()
    }

    override fun onClick(v: View) {
        if (v == mBinding.followNumSortBtn || v == mBinding.postNumSortBtn || v == mBinding.createTimeSortBtn || v == mBinding.commentNumSortBtn) {
            val order = when (v) {
                mBinding.followNumSortBtn -> { ORDER_FOLLOW_NUM }
                mBinding.postNumSortBtn -> { ORDER_THREAD_NUM }
                mBinding.createTimeSortBtn -> { ORDER_CREATE_TIME }
                mBinding.commentNumSortBtn -> { ORDER_COMMENT_NUM }
                else -> { ORDER_FOLLOW_NUM }
            }
            v.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
            mPage = 1
            mBinding.statusView.loading()
            collectionListAdapter.setNewData(ArrayList())
            mPresenter?.getCollectionList(mPage, mType, order)
        }
    }

    override fun setOnItemClickListener() {
        collectionListAdapter.setOnItemClickListener { adapter, view, position ->
            val intent = Intent(context, CollectionDetailActivity::class.java).apply {
                putExtra(Constant.IntentKey.COLLECTION_ID, collectionListAdapter.data[position].collectionId)
                putExtra(Constant.IntentKey.IS_NEW_CONTENT, collectionListAdapter.data[position].hasUnreadPost)
            }
            startActivity(intent)
        }

        collectionListAdapter.setOnItemChildClickListener { adapter, view, position ->
            if (view.id == R.id.latest_post_title) {
                val intent = Intent(context, NewPostDetailActivity::class.java).apply {
                    putExtra(Constant.IntentKey.TOPIC_ID, collectionListAdapter.data[position].latestPostId)
                }
                startActivity(intent)
            }
            if (view.id == R.id.avatar || view.id == R.id.author_name) {
                val intent = Intent(context, UserDetailActivity::class.java).apply {
                    putExtra(Constant.IntentKey.USER_ID, collectionListAdapter.data[position].authorId)
                }
                startActivity(intent)
            }
        }
    }

    override fun lazyLoad() {
        mPresenter?.getCollectionList(mPage, mType, mOrder)
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        mPage = 1
        mPresenter?.getCollectionList(mPage, mType, mOrder)
    }

    override fun onLoadMore(refreshLayout: RefreshLayout) {
        mPresenter?.getCollectionList(mPage, mType, mOrder)
    }

    override fun onGetCollectionListSuccess(collectionListBeans: List<CollectionListBean>, hasNext: Boolean) {
        mBinding.statusView.success()
        mBinding.refreshLayout.finishRefresh()

        if (mPage == 1) {
            if (collectionListBeans.isEmpty()) {
                mBinding.statusView.error("啊哦，这里空空的~")
            } else {
                collectionListAdapter.setNewData(collectionListBeans)
                mBinding.recyclerView.scheduleLayoutAnimation()
            }
        } else {
            collectionListAdapter.addData(collectionListBeans)
        }

        if (hasNext) {
            mPage ++
            mBinding.refreshLayout.finishLoadMore(true)
        } else {
            mBinding.refreshLayout.finishLoadMoreWithNoMoreData()
        }
    }

    override fun onGetCollectionListError(msg: String?) {
        mBinding.refreshLayout.finishRefresh()
        if (mPage == 1) {
            if (collectionListAdapter.data.size != 0) {
                showToast(msg, ToastType.TYPE_ERROR)
            } else {
                mBinding.statusView.error(msg)
            }
            mBinding.refreshLayout.finishLoadMore()
        } else {
            mBinding.refreshLayout.finishLoadMore(false)
        }
    }
}